tailnet/resume.go

```diff
@@ -0,0 +1,194 @@
+}
+
+type ResumeTokenProvider interface {
+	GenerateResumeToken(peerID uuid.UUID) (*proto.RefreshResumeTokenResponse, error)
+	VerifyResumeToken(token string) (uuid.UUID, error)
+}
+
+type ResumeTokenSigningKey [64]byte
+
+func GenerateResumeTokenSigningKey() (ResumeTokenSigningKey, error) {
+	var key ResumeTokenSigningKey
+	_, err := rand.Read(key[:])
+	if err != nil {
+		return key, xerrors.Errorf("generate random key: %w", err)
+	}
+	return key, nil
+}
+
+type ResumeTokenSigningKeyDatabaseStore interface {
+	GetCoordinatorResumeTokenSigningKey(ctx context.Context) (string, error)
+	UpsertCoordinatorResumeTokenSigningKey(ctx context.Context, key string) error
+}
+
+// ResumeTokenSigningKeyFromDatabase retrieves the coordinator resume token
+// signing key from the database. If the key is not found, a new key is
+// generated and inserted into the database.
+func ResumeTokenSigningKeyFromDatabase(ctx context.Context, db ResumeTokenSigningKeyDatabaseStore) (ResumeTokenSigningKey, error) {
+	var resumeTokenKey ResumeTokenSigningKey
+	resumeTokenKeyStr, err := db.GetCoordinatorResumeTokenSigningKey(ctx)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return resumeTokenKey, xerrors.Errorf("get coordinator resume token key: %w", err)
+	}
+	if decoded, err := hex.DecodeString(resumeTokenKeyStr); err != nil || len(decoded) != len(resumeTokenKey) {
+		newKey, err := GenerateResumeTokenSigningKey()
+		if err != nil {
+			return resumeTokenKey, xerrors.Errorf("generate fresh coordinator resume token key: %w", err)
+		}
+
+		resumeTokenKeyStr = hex.EncodeToString(newKey[:])
+		err = db.UpsertCoordinatorResumeTokenSigningKey(ctx, resumeTokenKeyStr)
+		if err != nil {
+			return resumeTokenKey, xerrors.Errorf("insert freshly generated coordinator resume token key to database: %w", err)
+		}
+	}
+
+	resumeTokenKeyBytes, err := hex.DecodeString(resumeTokenKeyStr)
+	if err != nil {
+		return resumeTokenKey, xerrors.Errorf("decode coordinator resume token key from database: %w", err)
+	}
+	if len(resumeTokenKeyBytes) != len(resumeTokenKey) {
+		return resumeTokenKey, xerrors.Errorf("coordinator resume token key in database is not the correct length, expect %d got %d", len(resumeTokenKey), len(resumeTokenKeyBytes))
+	}
+	copy(resumeTokenKey[:], resumeTokenKeyBytes)
+	if resumeTokenKey == [64]byte{} {
+		return resumeTokenKey, xerrors.Errorf("coordinator resume token key in database is empty")
+	}
+	return resumeTokenKey, nil
+}
+
+type ResumeTokenKeyProvider struct {
+	key    ResumeTokenSigningKey
+	clock  quartz.Clock
+	expiry time.Duration
+}
+
+func NewResumeTokenKeyProvider(key ResumeTokenSigningKey, clock quartz.Clock, expiry time.Duration) ResumeTokenProvider {
+	if expiry <= 0 {
+		expiry = DefaultResumeTokenExpiry
+	}
+	return ResumeTokenKeyProvider{
+		key:    key,
+		clock:  clock,
+		expiry: DefaultResumeTokenExpiry,
+	}
+}
+
+type resumeTokenPayload struct {
+	PeerID uuid.UUID `json:"sub"`
+	Expiry time.Time `json:"exp"`
```

Comment:
I'll make it store seconds since epoch but I won't use jwt.NumericDate because this code uses go-jose. Even though the NumericDate type would be compatible with this code, it feels wrong to import two different JWT packages in the same file.